### ISSUE TITLE
is it possible to embed in-game python scripting in pyglet?

### ISSUE THREAD

**Issue by mugenZebra** [pyglet/pyglet#280]
I was researching about this and saw a SO answer that says it might be possible, if it's already supported by pyglet, can you point to the doc that shows how? I couldn't find the doc on this after searching.

**Comment by GeeTransit** [comment]
What do you mean by in-game scripting? You can check out [`exec`](https://docs.python.org/3/library/functions.html#exec) and [`eval`](https://docs.python.org/3/library/functions.html#eval), but be very careful about how you use it.

**Comment by assyrianic** [comment]
why wouldn't it be possible? you'd likely have to create a Python plugin system though and an SDK for plugins to use.

**Comment by mugenZebra** [comment]
yeah thanks guys, I've experimented with exec and eval. I think now I just needs to find ways to limit the in-game Python usage to use limited language functionality or else the game will just be a mess of security problem lol

**Comment by GeeTransit** [comment]
tl;dr [Read this article by Ned Batchelder](https://nedbatchelder.com/blog/201206/eval_really_is_dangerous.html).

To be honest, restricting the language is pretty hard. You'd need to somehow go against what `exec` and `eval` are supposed to do, that is, running a string as if it was part of the source code.

I haven't tried this, but you could run the code with a special set of variables (by passing a dict of names and their values). Something like `exec(code, variables)` where `code` is the string that you are gonna execute and `variables` is the dict.

Another restriction would be checking and stopping code that uses double underscores (`__`) because they allow you to access some implementation specific stuff that probably won't be documented and can change frequently.

There are probably a few ways to bypass them, such as calling a method in the standard library that is always imported and always raises an exception (`codecs.Codec`) and getting the builtins from the exception, or getting double underscores with `getattr` and a string like `"\x5F\x5Fclass\x5F\x5F"` (which is `"__class__"`). I haven't thought of a way to get past both, but some dude out there's gonna find a way.

Some useful links:
https://nedbatchelder.com/blog/201206/eval_really_is_dangerous.html
https://www.reddit.com/r/Python/comments/hftnp/ask_rpython_recovering_cleared_globals/

**Comment by ryannathans** [comment]
you could sandbox something like micropython and build an interface if you want to give them access to certain things

**Comment by benmoran56** [comment]
I'll close this for now since it's more of a general Python thing, not specifically pyglet related. 

Thanks everyone.
